articles/modules/ROOT/pages/protecting-against-cypher-injection.adoc

```diff
@@ -38,119 +38,389 @@ Result result = session.run(queryString);
+Here is an example of another simple exploit input:
+
+`Input: ' RETURN a//`
 
-The parameter might get set like this:
 
 [source,cypher]
 ----
-:param name => "Robby' WITH DISTINCT true as haxxored MATCH (s:Student) DETACH DELETE s //"
+MATCH (s:Student) WHERE s.name = '' RETURN a//' RETURN s;
 ----
 
-And the parameterized query would look like this:
+Which results in the following database error:
+
+----
+
+Variable `a` not defined (line 1, column 44 (offset: 43))
+"MATCH (s:Student) WHERE s.name = '' RETURN a//' RETURN s;"
+                                            ^
+
+----
+
+If the server returns the raw error back, then the entire query is now visible, making it easier to send a more specific malicious input back.
+The attacker now knows the names of at least one label as well as the variables associated with it.
+
+To protect against this, aside from using parameters and sanitizing/validating user input, avoid returning database specific errors to the user and opt for more generic errors.
+
+Now we have discussed different types of Cypher Injection, so how do we protect against it?
+
+== Protecting against Cypher Injection using Parameters
+
+When receiving user input, it is possible to prevent Cypher Injection by using parameters.
+
+In the above example, instead of concatenating the students name with the `CREATE` query, one should use parameters instead:
+
+`:param name => "Robby' WITH DISTINCT true as haxxored MATCH (s:Student) DETACH DELETE s //"`
+
+The parameterized query would look like this:
+
 
 [source,cypher]
 ----
 CREATE (s:Student)
 SET s.name = $studentName
 ----
 
-When using parameters like this, it is impossible for the parameter, in part or in whole, to be interpreted as part of the query.
-It cannot hijack it.
+When using parameters like this, it is impossible for the parameter, in part of whole, to escape the query, stopping any attempts at Cypher Injection.
 
-One reason for this is that parameters are separate from the query. The query alone is compiled into an executable plan, and once compiled, can use any parameter map for execution.
-The parameters cannot alter the compiled plan, and have no opportunity to be included in the plan compilation.
+This is because parameters are separate from the query. The query alone is compiled into an executable plan, and once compiled, can use any parameter map for execution.
 
 In other words, once a query plan has been compiled, it is set, and nothing in the data submitted to it can change it, alter it, or hijack it.
 
-== Other common injection attempts also fail
+This type of protection is the easiest and safest way to protect against Cypher Injection on literals, unfortunately, not everything in a query is parameterizable.
 
-Not all inputs can be submitted as parameters.
-Maybe some malicious input made it into a CSV file for processing.
-A CSV of the names of new students for the year, for example.
+== Parameters and APOC
```

Comment:
This is the typo of info we will also be including the Security Guidance page in APOC. We can keep it in both areas, although you may ultimately wish to include a link to the APOC Guidance page.

Reply:
I think it is good to have here as well :) I agree about adding the link, we can do that at a later point! I'll make a note on the APOC guidance page ticket